Add video preprocessing (denoising) feature

t-sasatani opened the pull request:
This PR adds a denoising module for neural recording videos. It can be used offline, and with some updates, it could be used in real-time with the `streamDaq`.

It would be better to quickly merge this to the main branch to add other processing features in a relatively stable location. Also, nothing depends on this now anyway.

### Processing features
- Detect broken buffers by comparing buffers with the same position buffer in the previous frame (I'm only making mean error now, but we can also add more statistic functions).
- Gradient-based broken frame detection led by @MarcelMB. Merged from #94  
- Remove frames that have broken buffers. These broken frames are individually stacked and tracked to examine which frame got removed. ~Remove broken buffers by copying in the same position buffer in the previous frame.~
- Spatial frequency-based filtering.
- Generate the minimum projection out of a stack of frames.

Following is an example frame that detects noisy regions and patches it with the prior frame buffer:
![image](https://github.com/user-attachments/assets/397089ae-fec3-4731-9795-0005ec315934)

### Interface
You can run this with `mio process denoise` using an [example video](https://drive.google.com/open?id=1ELoWDditmomysVo4ad8xAJDS5cpWBBvU&usp=drive_fs).
The denoising parameters and export settings can be defined with [a YAML file](https://github.com/Aharoni-Lab/miniscope-io/blob/feat-preprocess/miniscope_io/data/config/process/denoise_example.yml).
```bash
mio process denoise -i .\user_dir\test.avi -c denoise_example
```

### Minor changes
- I moved `VideoWriter` to `io` module because it's pretty generic, and I wanted to use it for the denoising.

<!-- readthedocs-preview miniscope-io start -->
----
📚 Documentation preview 📚: https://miniscope-io--83.org.readthedocs.build/en/83/

<!-- readthedocs-preview miniscope-io end -->

sneakers-the-rat commented:
omfg you already did it!?!??!?!?! u are so fast. cant wait to check this out!!!!!

MarcelMB commented:
amazing!!!! Thanks, Takuya for working on this already!

- I was wondering if replacing a broken buffer with the one from the previous frame is a good strategy. Unsure. For one broken frame that might be fine but if there is a stretch of corrupted data and we basically have the same image for seconds not sure. 

- by subtracting a minimum projection at the end I was wondering if maybe not low-intensity biological signals could eventually get lost?

- I need to understand the frequency mask a bit more to see what goes through and what is blocked.

sneakers-the-rat commented:
> I was wondering if replacing a broken buffer with the one from the previous frame is a good strategy. Unsure. For one broken frame that might be fine but if there is a stretch of corrupted data and we basically have the same image for seconds not sure.

agree, I think we want to drop data rather than copy data - it's scientific data after all and the numbers matter, so duplicating a frame is effectively fabricating data (even though this is not a bad strategy if we were streaming a movie and didn't care about the pixel values as much)

> by subtracting a minimum projection at the end I was wondering if maybe not low-intensity biological signals could eventually get lost?
>
> I need to understand the frequency mask a bit more to see what goes through and what is blocked.

this is a decent pair of thoughts that sort of highlights that we might want to separate cosmetic/display-oriented processing stages from signal repair stages: the frequency filtering is repairing the image, while doing the minimum projection is just for display purposes (right? i'm assuming in analysis people will want to have the full signal). 

been working through backlog of issues after the holiday and my first big chunk of work will be on mio so i'll review this and stuff this week. thanks again for this Takuya :)

t-sasatani commented:
Thanks for the comments! I'm stuck with other stuff at the moment but will make it ready for review and ping you guys hopefully next week. Something I kind of want to know from @sneakers-the-rat is whether the high-level structure seems mergeable to the pipeline refactor. No intention to make this really compatible because we need this now and this is an independent module but the code structure is pretty arbitrary so it might be good if there is something I can easily anchor to.

And yes, Phil mentioned too dropping broken frames will be good so that's planned (might just add an option of dropping because masking is already there). I was just too lazy to look in how processing pipelines handle this and track dropped buffers.

For the FFT and minimum projection, these are just modules so I think these just should be available whenever needed. The frequency mask might be used in actual preprocessing before CMNE pipelines but I'm not sure.
https://github.com/Aharoni-Lab/Miniscope-v4/wiki/Removing-Horizontal-Noise-from-Recordings

MarcelMB commented:
adding a bug here for the buffer patching
most likely due to multiple frames being corrupted after each other and the code currently always looking at the last frame to patch the next one. Resulting in loosing most of each frame for the rest of the video

files used to get this bug: invivo-20241203-1-003, invivo-20241203-3.avi

using commit:
9cd2bc3955cd36d37771ad8ea5cc0dda9c02befb


![Screenshot 2025-01-09 at 12 49 12 PM](https://github.com/user-attachments/assets/a09c953d-710f-4036-ac5c-948d3e1a8847)

raymondwjang commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+    """
+
+    @staticmethod
+    def denoise(
```
seems like there's room for improvement in terms of SOLID, especially the [single responsibility principle](https://stackify.com/solid-design-principles/). This `denoise` method for example appears to:
1. check for module
2. read video
3. write video
4. among ~10 other functions that can/should probably be modularized.

The rule of thumb normally is that if there are more than 2-3 `if` statements in a non-c-style function, it's a symptom that it can be improved for readability and maintainability.  :)

I'd hazard a guess that a lot of these `if` blocks can be packaged into different classes and files with clearer names and more consistent formatting across methods for better recyclability.

sneakers-the-rat replied:
agreed. shouldn't be too tricky to split each of these into separate classes with an `__init__` that takes config, `process()` that applies the transform, and `finish()` (or whatever we want to call it) that handles writing output if requested.

t-sasatani replied:
Yeah, it's literally just a raw script disguised as some class now. I'll try to be more explicit that these aren't awaiting formatting review or avoid uploading PRs in draft status so we don't double labor on both ends.

sneakers-the-rat commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+
+            min_proj_frame = NamedFrame(name="min_proj", static_frame=minimum_projection)
+
+            if config.interactive_display.enable:
+                videos = [
+                    raw_video,
+                    noise_patch,
+                    patched_video,
+                    freq_filtered_video,
+                    freq_domain_video,
+                    min_proj_frame,
+                    freq_mask_frame,
+                ]
+                VideoPlotter.show_video_with_controls(
+                    videos,
+                    start_frame=config.interactive_display.start_frame,
+                    end_frame=config.interactive_display.end_frame,
+                )
+
```
here's a good example of why we want all these things to be separable functions/classes/etc. that can work framewise - ideally this would work on demand, where we just open the UI and compute it as requested in the context of a GUI (not something `mio` should try and provide) rather than needing to precompute everything and set up a UI for it afterwards.

t-sasatani replied:
Yeah makes sense. This will be mainly for debugging, as the interactive plot gets very slow after a few hundred frames. I guess I'll put non-user facing stuff in the debug section of the config.

sneakers-the-rat commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+        return min_projection
+
+    @staticmethod
+    def normalize_video_stack(image_list: list[np.ndarray]) -> list[np.ndarray]:
```
either should be a separate class like the other processor classes or a function, strange to have it here in a different form than all the other video processing methods.

t-sasatani replied:
Done

sneakers-the-rat reviewed (changes requested):
Alright, thanks for patience, finally got time to review this. 

Two changes I think we should for sure make before merging this:
- split out each of the processing methods into separate classes with a common API and then make the video processor class generic over those classes, details in comments. 
- unify the existing `Frame` and `Frames` classes with the new `NamedFrame` class, making separate classes for a single frame vs. collections of frames.
- have tests for all of this

The rest are perf improvements and style comments

raymondwjang commented:
Also would like to add that while I'm also sometimes a violator of this, each PR should ideally consist of ~10 commits, with fewer than 500 lines of codes.

@t-sasatani: Here's something that might help that @sneakers-the-rat and I presented during the lab meeting that you couldn't attend [(link)](https://docs.google.com/presentation/d/16ZRpjDD3BZndC2VmfzuZ4AtqW1c8TaRperaaUXD0MyQ/edit#slide=id.g3164360f84a_0_131)

t-sasatani commented:
Thanks. I'll look at these later, but could you guys hold off on formatting reviews while the PR is marked draft? I commented in this thread last week that I'll change it to ready for review and ping you. It's not like I'm bothered or anything, but I do prefer using the draft as draft so we can discuss experimental prototypes before thinking about formatting or structure.

t-sasatani commented:
And thanks for the info, @raymondwjang. I think I attended that meeting, but I sent a view request because I couldn't access the slides.

I can squash commit counts later if you prefer, but I guess it is unlikely that PRs for refactoring or adding new modules (like this one) will fit in 500 lines of code. That rule makes sense for minor additions or fixes, but is it considered better for everything?

raymondwjang commented:
shared! obviously there are reasonable violations to the rule, but simply seeing 1000+ lines on a PR is pretty threatening as a reviewer. 

even within enterprise level large projects, massive PRs are rare unless there's a sweeping feature change or a massive refactor, as they always introduce exponentially larger review overheads, bug risks, rollback costs, etc. Even with new files it not super common to see PRs approaching 1000 and above, and that's including the tests and documentations. Also remember that these repos are generated hosted with an extensive suite of CI/CD tools that make sure the codes are healthy. (which we don't currently have)

currently this code is 1000 lines, but with tests and documentations it might approach 2000. that can turn into hours of reviewing and still missing bugs and edge cases. i had a similar pr that grew uncontrollably, and ended up having to chop it up into 5 different PRs so that it's somewhat reviewable.

as @sneakers-the-rat can attest, i often fail to keep things nicely packaged for reviewers and future me as well. but from what i've learned, it seems like it's always recommended to keep things small, nimble and modular. do as i say not as i do lmao

sneakers-the-rat commented:
>  I commented in this thread last week that I'll change it to ready for review and ping you. It's not like I'm bothered or anything, but I do prefer using the draft as draft so we can discuss experimental prototypes before thinking about formatting or structure.

no problem! i was just getting started working on mio again and wanted to catch up here, draft status totally understood, giving me feedback that hopefully might help with some ideas on different ways to finalize it :)

and Raymond super appreciate the advocacy of small PRs. There are a few gigantic refactors that need to get done so i'll be guilty of a few very large PRs myself here soon enough, but once we get into a roughly stable code structure we should be able to iterate much more cleanly :).

t-sasatani commented:
Ok I'll try to keep these short, though I'll probably violate it sometimes. It might be easier for me to overlook review comments for draft PRs, and sry for that in advance, but if you're ok with that, I'm fine.

t-sasatani commented on .gitignore:
```diff
@@ -159,4 +159,8 @@ cython_debug/
-.pdm-python
\ No newline at end of file
+.pdm-python
+user_dir/*
```
You're talking about the user_dir? I just wanted somewhere to keep data while I work in the repo for my convenience. I'll get rid of them before getting out of draft state.

t-sasatani replied:
Oh I forgot user_dir is used for config. I might keep a user_data dir just for convenience in dev.

sneakers-the-rat replied:
up to you where you want to put it, no problem having an extra line in .gitignore. 

if you wanted this to persist you would set `user_dir` in the global config for your machine (`mio config global path`)

t-sasatani commented on .gitignore:
```diff
@@ -159,4 +159,8 @@ cython_debug/
+user_dir/*
+!user_dir/.gitkeep
+
+~/.config/miniscope_io/logs/
```
Oh this is just an accident same as above

t-sasatani commented on mio/data/config/process/denoise_example.yml:
```diff
@@ -0,0 +1,29 @@
+id: denoise_example
+mio_model: tests.test_mixins.LoaderModel
+mio_version: 0.6.1.dev16+g6436633.d20241211
```
Maybe something upstream in the config PR? We can probably keep this and adjust it later after the config stuff is fixed?

t-sasatani commented on mio/io.py:
```diff
@@ -19,6 +19,107 @@
 from mio.types import ConfigSource
 
 
+class VideoWriter:
```
Yeah just pulled this out from stream daq without thinking much hre

t-sasatani commented on mio/io.py:
```diff
@@ -19,6 +19,107 @@
+        """
+        self.video_path = video_path
+        self.cap = cv2.VideoCapture(str(video_path))
+        self.height = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+        self.width = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
```
If we want flexibility here, that makes sense, but as we probably don't want it wouldn't that just be harmful? I was thinking of one instance per video path here.

sneakers-the-rat replied:
yeah that should be the expected pattern, but just general "best practices" object oriented design, it is *possible* and conceivable that someone would manually change `video_path`, and doing stuff with properties is one way of handling state changes like that - always read from the current thing by programming dependencies between attributes. i would say comparatively minor so don't worry abt it

t-sasatani replied:
Switched to @property

t-sasatani commented on mio/models/frames.py:
```diff
@@ -0,0 +1,136 @@
+        None,
+        description="Frame data, if provided.",
+    )
+    video_frame: Optional[List[np.ndarray]] = Field(
```
Yeah that side of the repo is a mystery to me and we can leave that for a pure refactor PR

t-sasatani replied:
Split out classes to make it bit more explicit

t-sasatani commented on mio/models/frames.py:
```diff
@@ -0,0 +1,136 @@
+        ...,
+        description="Name of the video.",
+    )
+    static_frame: Optional[np.ndarray] = Field(
+        None,
+        description="Frame data, if provided.",
+    )
+    video_frame: Optional[List[np.ndarray]] = Field(
+        None,
+        description="Video data, if provided.",
+    )
+    video_list_frame: Optional[List[List[np.ndarray]]] = Field(
+        None,
+        description="List of video data, if provided.",
+    )
```
Oh yeah, I was wondering how to handle truncated videos and distinguish a stack of frames coming from truncated videos and a regular video clip. This is something I was going to clean up by just making separate models for these.

t-sasatani replied:
Just made separate models for these

t-sasatani commented on mio/models/frames.py:
```diff
@@ -0,0 +1,136 @@
+        None,
+        description="List of video data, if provided.",
+    )
+    frame_type: Optional[str] = Field(
```
as the above just truncated videos, something that just grew while sketching out and awaiting cleanum

t-sasatani replied:
Got rid of this

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,142 @@
+
+class NoisePatchConfig(BaseModel):
+    """
+    Configuration for patch based noise handling.
```
Yup palnned

t-sasatani replied:
Done

t-sasatani commented on tests/test_models/test_model_process.py:
```diff
@@ -0,0 +1,33 @@
+from ..conftest import DATA_DIR, CONFIG_DIR
+
+def test_interactive_display_config():
+    config = DenoiseConfig.from_id("denoise_test").interactive_display
```
@sneakers-the-rat, it might be in docs somewhere, but could you let me know how to find these configurations in the test dir? I'm putting it into the same directory as the test config for stream DAQ, but getting this. Do you need some registering or anything?

>No config with id denoise_test found in /path/to/somewhere/Application Support/mio/config

t-sasatani replied:
Just leaving a note here so I don't forget I'm skipping these config tests for now.

sneakers-the-rat replied:
In general I don't think we need these kinda tests (if we wanted to ensure that these values don't change, we could just copy it to tests and assert that the copy in tests was equal to the one in the package), but the fact that it isn't being found would be a bug with the config discovery methods or how we are monkeypatching them to include the `tests/data/config` directory during testing. i'll check this out in the morning

t-sasatani commented:
I'm switching this to ready for review as I'm heading home. I'm sorry it's becoming gigantic.

It's done structure-wise, and I changed the patch frame function to drop frames and added a tracking/export thing.

\I haven't gone through the reviews yet, more documents/tests need to be completed, and the interfaces, like the naming of the export paths, are unsettled, so I'll return to them later this week. If you want to take over or make a significant change in the meantime, feel free to go ahead

sneakers-the-rat commented on mio/process/frame_helper.py:
```diff
@@ -0,0 +1,280 @@
+        return mask
+
+
+class ZStackHelper:
```
i like this idea :)

sneakers-the-rat commented on mio/process/frame_helper.py:
```diff
@@ -0,0 +1,280 @@
+        self.height = height
+        self.width = width
+
+    def split_by_length(self, array: np.ndarray, segment_length: int) -> list[np.ndarray]:
```
something like this seems pretty general purpose, but we don't have a good place to put 'general array manipulation functions' like a utils junk drawer would be. i like this helper class idea, bundling some operations together with some staticmethods. i get where this is coming from being split off from the previous implementation, so i'm fine with this, but in the future i might want to play with this helper class idea, make it its own module, and make some generic "ArrayHelper"/"BufferHelper" kind of classes

t-sasatani replied:
Yeah, that does make sense. After all, this whole package is about tinkering with numpy arrays not just this denoising module.

t-sasatani commented on mio/io.py:
```diff
@@ -19,6 +19,107 @@
+
+            yield index, frame
+
+    def release(self) -> None:
```
Yea this is fragile, as you say. Switched to property

t-sasatani commented on mio/models/frames.py:
```diff
@@ -0,0 +1,136 @@
+from mio.io import VideoWriter
+from mio.logging import init_logger
+
+T = TypeVar("T", np.ndarray, List[np.ndarray], List[List[np.ndarray]])
```
Just split out video and frame classes here because it's not really necessary to unify them.

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,142 @@
+    method: str = Field(
+        default="mean_error",
+        description="Method for handling noise.",
+    )
```
Seems like it was done in one of the attached PRs

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,142 @@
+        default=20,
+        description="Threshold for detecting noise.",
+    )
+    buffer_size: int = Field(
+        default=5032,
+        description="Size of the buffers composing the image."
+        "This premises that the noisy area will appear in units of buffer_size.",
+    )
```
The line-by-line approach seems simpler, so let's make that the default. I will leave this actual buffer-wise comparison approach as a non-recommended option because it will likely be more accurate.

t-sasatani replied:
I changed it a bit so this gets calculated from the device config that acquired this stream.

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,142 @@
+        default=None,
+        description="Frame to end processing at.",
+    )
+    output_result: bool = Field(
+        default=True,
+        description="Whether to output the result.",
+    )
+    output_dir: Optional[str] = Field(
+        default=None,
+        description="Directory to save the output in.",
+    )
```
Got it. I will skip it here.

t-sasatani commented on mio/plots/video.py:
```diff
@@ -0,0 +1,140 @@
+
+    @staticmethod
+    def show_video_with_controls(
+        videos: List[NamedFrame],
```
Done by splitting out NamedFrame and Video classes

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+        Returns:
+        FrameProcessor: A FrameProcessor object.
+        """
+        self.height = height
```
I guess everything except for the spacial freq mask can be done that way. Will switch.

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+    plt = None
+
+
+class FrameProcessor:
```
I guess it's now kind of like this.

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+        self.height = height
+        self.width = width
+
+    def split_by_length(self, array: np.ndarray, segment_length: int) -> list[np.ndarray]:
```
I didn't notice that `numpy.split` can do this kind of split with 1-D array indexes. Will switch.

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+        buffer_has_noise = False
+        for buffer_index in range(buffer_per_frame):
+            for split_index in range(noise_patch_config.buffer_split):
+                i = buffer_index * noise_patch_config.buffer_split + split_index
```
Yeah this is confusing in general having splitting and processing mixed. Will change structure here.

t-sasatani replied:
Done

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+                if cv2.waitKey(1) == 27:  # Press 'Esc' key to exit visualization
+                    break
+            cv2.destroyAllWindows()
+        return mask
```
Yeah, moved to the model. I think we might want some interactive display option before the actual processing happens.

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+            diff_frames = []
+        if config.frequency_masking.enable:
+            freq_domain_frames = []
+            freq_filtered_frames = []
```
I think it's kind of like this now

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+
+            subtract_minimum = [(frame - minimum_projection) for frame in output_frames]
+
+            subtract_minimum = VideoProcessor.normalize_video_stack(subtract_minimum)
```
Done

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+
+                if config.frequency_masking.enable:
+                    freq_filtered_frame, frame_freq_domain = processor.apply_freq_mask(
+                        img=patched_frame,
```
I wish I could pass down pointers here. It's slightly better now, but it still uses extra memory if the processing isn't enabled (and, of course, it is pretty inefficient inside the helpers).

t-sasatani replied:
Well at least the input isn't copied now so will resolve it

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+                    freq_filtered_frames.append(freq_filtered_frame)
+                    output_frame = freq_filtered_frame
+                output_frames.append(output_frame)
+        finally:
```
I think it's now like this. It might be better to release memory if it gets more integrated with other stuff, but I guess just having export stuff is fine while this is an independent module.

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,142 @@
+        default=1,
+        description="Multiplier for the difference between the mean and the pixel value.",
+    )
+    output_result: bool = Field(
+        default=False,
+        description="Whether to output the result.",
+    )
+    output_noise_patch: bool = Field(
+        default=False,
+        description="Whether to output the noise patch.",
+    )
+    output_diff: bool = Field(
+        default=False,
+        description="Whether to output the difference.",
+    )
```
done

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,142 @@
+    )
+    spatial_LPF_cutoff_radius: int = Field(
+        default=...,
+        description="Radius for the spatial cutoff.",
```
added

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,142 @@
+    )
+    end_frame: Optional[int] = Field(
+        default=None,
+        description="Frame to end processing at.",
```
added

t-sasatani commented on mio/plots/video.py:
```diff
@@ -0,0 +1,140 @@
+        fig, axes = plt.subplots(1, num_streams, figsize=(20, 5))
+
+        frame_displays = []
+        for idx, ax in enumerate(axes):
+            initial_frame = video_frames[idx][0]
+            frame_display = ax.imshow(initial_frame, cmap="gray", vmin=0, vmax=255)
+            frame_displays.append(frame_display)
+            if titles:
+                ax.set_title(titles[idx])
+            ax.axis("off")
```
Haven't brought it to the level it can be used with streamdaq but at least chunked it up in units that hopefully make sense.

t-sasatani commented on mio/plots/video.py:
```diff
@@ -0,0 +1,140 @@
+        num_frames = max(len(stream) for stream in video_frames)
+        logger.info(f"Max stream length: {num_frames}")
+
+        fig, axes = plt.subplots(1, num_streams, figsize=(20, 5))
```
Not exactly sure what this was because it's pretty outdated, but it doesn't have to be actually interactive and doing `draw_idle()` whenever GUI gets released seems fine.

t-sasatani commented on mio/plots/video.py:
```diff
@@ -0,0 +1,140 @@
+    """
+
+    @staticmethod
+    def show_video_with_controls(
```
Yes done this

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+
+        buffer_has_noise = False
+        for buffer_index in range(buffer_per_frame):
+            for split_index in range(noise_patch_config.buffer_split):
```
I fixed the code to look at pixel chunks that accumulate from the end of the buffer (i.e., how the error is likely to happen). But anyway, the mean_error approach is unstable, and I posted a warning not to use it.

t-sasatani commented on mio/process/video.py:
```diff
@@ -0,0 +1,429 @@
+
+        return sub_arrays
+
+    def patch_noisy_buffer(
```
I think I'll leave this for later as the mean_error approach won't be used for now.

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,142 @@
+    display_mask: bool = Field(
+        default=False,
+        description="Whether to display the mask.",
+    )
```
Done

t-sasatani commented:
Thanks for all the feedback. I might have missed something, but I went through all the comments. 
- We should make `gradient` the primary noise detection approach, but I am also thinking about leaving the `mean_error` method in case (probably until we start benchmarking stuff and combining methods). This does make the helper methods slightly more complex so let me know if I should move this to a different branch or something.
- I think the YAML interface and export path need improvement, but that's pretty minor, so I think we can push that later.

t-sasatani commented:
Just failing at coverage reports. Probably because Coveralls is down for maintenance.

sneakers-the-rat commented on tests/test_models/test_model_frames.py:
```diff
@@ -0,0 +1,60 @@
+        # Ensure that no write methods are called
+        mock_VideoWriter.write_frame.assert_not_called()
+        mock_VideoWriter.assert_not_called()
+
+if __name__ == '__main__':
+    unittest.main()
```
```suggestion

```

sneakers-the-rat commented on tests/test_plots/test_plots_video.py:
```diff
@@ -0,0 +1,49 @@
+    @patch('mio.plots.video.plt.subplots')
+    @patch('mio.plots.video.Slider')
+    @patch('mio.plots.video.Button')
+    def test_show_video_with_controls(self, MockButton, MockSlider, mock_subplots, mock_show):
```
this doesn't *really* test anything important about this class, but since it's not mission critical and something we will take out and put in a separate GUI package anyway, fine for now.

t-sasatani replied:
Makes sense. Not exactly sure how to test widgets at the moment, but can do it later if we're isolating this.

sneakers-the-rat commented on tests/test_plots/test_plots_video.py:
```diff
@@ -0,0 +1,49 @@
+        mock_subplots.assert_called_once()
+
+        # Verify the number of axes matches the number of videos
+        self.assertEqual(len(mock_axes), len(self.named_frames))
```
classic LLM assertion that tests nothing

t-sasatani replied:
Yeah I guess the called times is the only ones that make sense here

sneakers-the-rat commented on tests/test_plots/test_plots_video.py:
```diff
@@ -0,0 +1,49 @@
+        # Verify that plt.show() is called to display the plot
+        mock_show.assert_called_once()
+
+if __name__ == '__main__':
+    unittest.main()
```
```suggestion

```

sneakers-the-rat commented on tests/test_process_video.py:
```diff
@@ -0,0 +1,62 @@
+
+class TestVideoProcessors(unittest.TestCase):
+    def setUp(self):
+        self.test_dir = Path("/tmp/test_output")
```
not portable across OS, unfortunately, and likely to hit some existing directory this way. this is why we use the `tmp_path` pytest fixture https://docs.pytest.org/en/latest/how-to/tmp_path.html

sneakers-the-rat replied:
this would also need to get removed on teardown

sneakers-the-rat replied:
i'm actually just going to rewrite these with pytest

t-sasatani replied:
Nice thanks!

sneakers-the-rat commented on tests/test_stream_daq.py:
```diff
@@ -41,11 +41,7 @@ def default_streamdaq(set_okdev_input, request) -> StreamDaq:
-                "3676bc4c6900bc9ec18b8387abdbed35978ebc48408de7b1692959037bc6274d",
-                "3891091fd2c1c59b970e7a89951aeade8ae4eea5627bee860569a481bfea39b7",
-                "d8e519c1d7e74cdebc39f11bb5c7e189011f025410a0746af7aa34bdb2e72e8e",
+                "ee7bdb97c1e98ebeefc65ae651968e3a72d099e57d1fdec5ec05a3598733db93",
```
hell yes.

sneakers-the-rat reviewed (commented):
nice. ok i'm good with all this except for one thing in the tests which will cause us trouble down the line, give me one second to update it, then i'll throw it back to @t-sasatani , and then after resolving last questions we should be good to go

sneakers-the-rat commented:
ok this is just about my minimum for "is this PR tested" - i think all the core behavior has *basic* tests, but there is a bunch of accessory code that is untested - like the plotting widget, the wrappers around all the intermediate processing stages, their outputs, etc. but i'm fine with it for now since a lot of that is just wrapping the core behavior in `process/frame_helpers`

the thing that I don't really love is that we now have two sets of models for images and frames: `models.data` and `models.frames` - that shouldn't last long, and i'll raise a followup issue for that (raised #118)

t-sasatani commented on mio/io.py:
```diff
@@ -19,6 +20,124 @@
+                "-r": str(fps),
+            }
+        else:
+            output_dict["-r"] = str(fps)
```
Totally makes sense

t-sasatani commented on mio/models/frames.py:
```diff
@@ -0,0 +1,125 @@
+        description="Name of the video.",
+    )
+    model_config = ConfigDict(
+        arbitrary_types_allowed=True,
```
Yeah this is pretty annoying. Definitely next PR!

t-sasatani commented on mio/models/frames.py:
```diff
@@ -0,0 +1,125 @@
+
+    def display(self, binary: bool = False) -> None:
+        """
+        Display the frame data.
```
Yes exactly, will do.

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,258 @@
+
+    threshold: float = Field(
+        ...,
+        description="Threshold for detecting invalid frames based on mean squared error.",
```
Just raw mean square error (not normalized), so it's any positive number or zero.

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,258 @@
+        "This is used in the mean_error method to compare frames"
+        " in the units of data transfer buffers.",
+    )
+    buffer_size: int = Field(
```
Good catch. I changed it based on the device config already, but forgot to delete this field. Just deleted now

t-sasatani commented on mio/models/process.py:
```diff
@@ -0,0 +1,258 @@
+    black_area_config: Optional[BlackAreaDetectorConfig] = Field(
+        default=None,
+        description="Configuration for detecting invalid frames based on black area.",
+    )
```
Actually, these caused problems. I added some tests to bind that.

sneakers-the-rat commented on tests/test_process/test_video.py:
```diff
@@ -37,6 +43,16 @@ def test_noise_patch_processor(video_frame, tmp_path):
+
+    processor = NoisePatchProcessor("denoise_example", denoise_config.noise_patch, tmp_path)
+    processed_frame = processor.process_frame(random_8bit_video_frame)
+    assert processed_frame is random_8bit_video_frame
```
thx for this

t-sasatani commented:
Sure! Thanks for looking into it again.
